src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -6542,6 +6542,23 @@ IArrayInitializerOperation popAndAssembleArrayInitializerValues(IArrayInitialize
         public override IOperation? VisitCollectionExpression(ICollectionExpressionOperation operation, int? argument)
         {
             EvalStackFrame frame = PushStackFrame();
+
+            if (operation.ConstructArguments.Any(a => a is IArgumentOperation) && !operation.ConstructArguments.All(a => a is IArgumentOperation))
+                throw ExceptionUtilities.UnexpectedValue("Mixed argument operations and non-argument operations in ConstructArguments");
+
+            // Ugly, but necessary.  If we bound successfully, we'll have an array of IArgumentOperation.  We want to
```

Comment:
>Ugly, but necessary.   [](http://example.com/codeflow?start=15&length=22)

I do not find this statement useful in general, and, frankly, I do not find the code ugly. We are always trying to do what is necessary to have correct behavior for all possible scenarios.